modules/linear_boltzmann_solvers/executors/pi_keigen_smm.cc

```diff
@@ -0,0 +1,1212 @@
+      } // for face
+    }   // for cell
+    ++gs;
+  } // for groupset
+
+  // Set the ghosted tensor vector with the local tensors, then
+  // communicate the ghosts
+  tensors_->Set(local_tensors);
+  tensors_->CommunicateGhostEntries();
+}
+
+std::vector<double>
+PowerIterationKEigenSMM::ComputeSourceCorrection() const
+{
+  const auto& grid = lbs_solver_.Grid();
+  const auto& pwld = lbs_solver_.SpatialDiscretization();
+  const auto& matid_to_xs_map = lbs_solver_.GetMatID2XSMap();
+  const auto& unit_cell_matrices = lbs_solver_.GetUnitCellMatrices();
+
+  const auto& diff_sd = diffusion_solver_->SpatialDiscretization();
+  const auto& diff_uk_man = diffusion_solver_->UnknownStructure();
+
+  const auto first_grp = front_gs_.groups_.front().id_;
+  const auto num_gs_groups = front_gs_.groups_.size();
+
+  auto tensors = tensors_->MakeGhostedLocalVector();
+
+  // Create the output vector
+  ParallelSTLVector output(
+    diff_sd.GetNumLocalDOFs(diff_uk_man), diff_sd.GetNumGlobalDOFs(diff_uk_man), mpi_comm);
+
+  // Build the source
+  for (const auto& cell : grid.local_cells)
+  {
+    const auto& rho = lbs_solver_.DensitiesLocal()[cell.local_id_];
+    const auto& cell_mapping = pwld.GetCellMapping(cell);
+    const auto nodes = cell_mapping.GetNodeLocations();
+    const auto num_cell_nodes = cell_mapping.NumNodes();
+    const auto num_cell_faces = cell.faces_.size();
+
+    const auto& fe_values = unit_cell_matrices[cell.local_id_];
+    const auto& K = K_tensor_matrices_[cell.local_id_];
+
+    const auto& xs = matid_to_xs_map.at(cell.material_id_);
+    const auto& sigma_tr = xs->SigmaTransport();
+
+    // Volumetric term
+    for (int gsg = 0; gsg < num_gs_groups; ++gsg)
+    {
+      const auto g = first_grp + gsg;
+      const auto& sig_tr = rho * sigma_tr[g];
+
+      // Compute the volumetric correction term to the diffusion source
+      // for this diffusion DoF. Unlike in many other areas of the codebase,
+      // these tensors are (dim x dim) instead of (3 x 3) so no special
+      // treatment is needed for the 1D case when only the z-component is
+      // present. The term is given by:
+      // grad bi dot div(bj Tj) / sigma_tr
+
+      for (int i = 0; i < num_cell_nodes; ++i)
+      {
+        const auto imap = diff_sd.MapDOF(cell, i, diff_uk_man, 0, gsg);
+
+        double val = 0.0;
+        for (int j = 0; j < num_cell_nodes; ++j)
+        {
+          // Here, because the tensors are transport-based, the transport
+          // discretization is used to obtain the tensor DoF.
+          const auto jmap = pwld.MapDOFLocal(cell, j, tensor_uk_man_, g, 0);
+          const double* T = &tensors[jmap];
+
+          // This adds the rank-2 tensor contraction (double dot product)
+          // of the closure tensor and the tensor stiffness matrix for nodes
+          // i and j.
+          for (int k = 0; k < dimension_; ++k)
+            for (int l = 0; l < dimension_; ++l)
+              val += *T++ * K[i][j][k][l];
```

Comment:
We shouldn't do this: `*T++`. I'll just leave the one comment, but it applies to several places in this file.

Reply:
Just out of curiosity, why should we not do this? Readability? I have always liked doing this for contiguous data since it I don't like having to index like `T[k * dimension_ + l]`, or `psi[cell_id * num_cell_nodes * num_dirs * num_grps + i * num_dirs * num_grps + d * num_grps + g]` when taken to more of an extreme.

Reply:
I think we should avoid pointer arithmetic wherever possible. It may be more concise, but it's also error prone. Even though it looks messier, I think it's easier to see what's happening when you use index notation - especially for new developers just learning the code and for developers having to maintain the code.

Reply:
Yea, that's fair. It is self documenting, but very gross, haha! Making the change. We should probably discuss this on Wednesday and create an issue for removing pointer arithmetic if we decide that's the route we want to go. I've always thought that `*T++` was faster than `T[k * dimension_ + l]`. Is that not the case? Or is it inconsequential? I'll make the change.

Reply:
I strongly suspect that there is no difference in performance between the two. The index notation is going to end up being pointer arithmetic after the compiler is finished with it.